science/platform-implementations/scd40.cpp

```diff
@@ -0,0 +1,129 @@
+// #pragma once
+
+#include "scd40.hpp"
+
+using namespace std::chrono_literals;
+using scd40_nm = sjsu::science::scd40;
+
+
+scd40_nm::scd40(hal::i2c& p_i2c, hal::steady_clock& p_clock) : m_i2c(p_i2c), m_clock(p_clock) {}
+
+hal::result<scd40_nm> scd40_nm::create(hal::i2c& p_i2c, hal::steady_clock& clock){
+    scd40 scd40(p_i2c, clock);
+    HAL_CHECK(scd40.start());
+    return scd40;
+
+}
+
+hal::status scd40_nm::start(){
+    std::array<hal::byte, 2> start_address =  { start_periodic_measurement_first_half, start_periodic_measurement_second_half };
+    HAL_CHECK(hal::write(m_i2c, addresses::device_address, start_address, hal::never_timeout()));
+    return hal::success();
+}
+
+hal::result<scd40_nm::scd40_read_data> scd40_nm::read() {
+    std::array<hal::byte, 2> read_address = {read_measurement_first_half, read_measurement_second_half };
+    std::array<hal::byte, 9> buffer;
+
+    HAL_CHECK(hal::write(m_i2c,addresses::device_address, read_address));
+    hal::delay(m_clock, 1ms);
+    HAL_CHECK(hal::read(m_i2c, addresses::device_address, buffer, hal::never_timeout()));
+    
+    scd40_nm::scd40_read_data rd;
+    rd.co2 = buffer[0] << 8 | buffer[1];
+    rd.temp = (-45 + 175.0*(buffer[3] << 8 | buffer[4])/ (1 << 16));
+    rd.rh = 100.0 * (buffer[6] << 8 | buffer[7]) / (1 << 16);
+
+    return rd;
+}
+
+hal::status scd40_nm::stop() {
+    std::array<hal::byte, 2> stop_address =  { stop_periodic_measurement_first_half, stop_periodic_measurement_second_half };
+    HAL_CHECK(hal::write(m_i2c, addresses::device_address, stop_address, hal::never_timeout()));
+    return hal::success();
+}
+
+hal::result<scd40_nm::scd40_settings> scd40_nm::get_settings() {
+    std::array<hal::byte, 2> read_address_temp = {get_temperature_offset_first_half, get_temperature_offset_second_half };
+    std::array<hal::byte, 2> read_address_alt = { get_sensor_altitude_first_half, get_sensor_altitude_second_half };
+    std::array<hal::byte, 6> buffer;
+
+    HAL_CHECK(hal::write(m_i2c,addresses::device_address, read_address_temp));
+    hal::delay(m_clock, 1ms);
+    HAL_CHECK(hal::read(m_i2c, addresses::device_address, buffer, hal::never_timeout()));
+    
+    scd40_nm::scd40_settings get;
+     
+    get.temp_offset = 175 *(buffer[0] << 8 | buffer[1])/(1 << 16);
+
+    HAL_CHECK(hal::write(m_i2c,addresses::device_address, read_address_alt));
+    hal::delay(m_clock, 1ms);
+    HAL_CHECK(hal::read(m_i2c, addresses::device_address, buffer, hal::never_timeout()));
+    get.altitude = buffer[0]<<8 | buffer[1];
+
+    return get;
+}
+
+hal::status scd40_nm::set_settings( struct settings setting) {
+    if(setting.set_temp != 0){
+        int temp = int(((setting.set_temp*(1<<16)) / 175)) ;
```

Comment:
```suggestion
        int temp = int(((setting.set_temp*(1<<16)) / 175)) ;
```
Give these magic numbers names. 175 isn't some power of two so it must mean something. Fine the term in the datasheet and make a constexpr variables for it. Something like:

```C++
// Found on page X in datasheet name "<url to datasheet>"
constexpr std::uint32_t temperature_scalar = 175; 
```